nlp/language_modeling/megatron_gpt_prune.py

```diff
@@ -0,0 +1,149 @@
+
+
+@hydra_runner(config_path="conf", config_name="megatron_gpt_prune")
+def main(cfg) -> None:
+    if not torch.cuda.is_available():
+        raise EnvironmentError("GPU is required for the pruning.")
+
+    # Overwrite model config with the one from the model checkpoint and apply pruning modifications
+    model_cfg = load_config(cfg.model.restore_from_path)
+    model_cfg.update(cfg.model)
+    model_cfg.name = "modelopt"
+
+    trainer = Trainer(strategy=NLPDDPStrategy(), **cfg.trainer)
+    model = MegatronGPTModel.restore_from(
+        restore_path=cfg.model.restore_from_path, override_config_path=model_cfg, trainer=trainer
+    )
+    model.freeze()
+
+    data_iter = get_calib_data_iter(
+        cfg.prune.calib_dataset,
+        cfg.inference.batch_size,
+        cfg.prune.num_calib_size,
+        cfg.inference.max_context_length,
+    )
+    dataloader = [data for data in data_iter]
+
+    def forward_loop(model):
+        # NOTE: Alternatively you can also use `model.forward_bwd_step(data_iter, forward_only=True)`
+        # if your model is setup for training.
+        model.set_inference_config(OmegaConf.to_container(cfg.inference))
+        for i, batch in enumerate(tqdm(dataloader, desc="Calibrating")):
+            model.predict_step(batch, i)
+
+    for name, param in model.state_dict().items():
+        if param is not None:
+            print_rank_0(f"model params before prune: {name} -> {param.type()} {param.shape}")
+
+    model_pruned, _ = mtp.prune(
+        model,
+        mode="mcore_gpt_minitron",
+        constraints={
+            "export_config": {
+                "ffn_hidden_size": cfg.prune.ffn_hidden_size,
+                "num_attention_heads": cfg.prune.num_attention_heads,
+                "num_query_groups": cfg.prune.num_query_groups,
+            }
+        },
+        dummy_input=None,  # Not used
+        config={
+            "forward_loop": forward_loop,
+        },
+    )
+
+    del model
+
+    for name, param in model_pruned.state_dict().items():
+        if param is not None:
+            print_rank_0(f"model params after prune: {name} -> {param.type()} {param.shape}")
+
+    # NOTE: Here we correct the pruned model configs here because the current API does not yet correct the configs internally
+    # TODO: remove them once the API internally correct the configs
+
+    kv_channels = model_cfg["kv_channels"] if "kv_channels" in model_cfg else None
+    kv_channels = kv_channels if kv_channels is not None else model_cfg.hidden_size // model_cfg.num_attention_heads
+
+    model_pruned.cfg["ffn_hidden_size"] = cfg.prune.ffn_hidden_size
+    model_pruned.cfg["num_attention_heads"] = cfg.prune.num_attention_heads
+    model_pruned.cfg["num_query_groups"] = cfg.prune.num_query_groups
+    model_pruned.cfg["kv_channels"] = kv_channels
+
+    model_pruned.save_to(cfg.export.save_path)
+
+    torch.distributed.barrier()
+
+    # TODO: read back and run inference
+
+
+if __name__ == '__main__':
+    main()
```

Comment:
Cleaned up your code

````python
import modelopt.torch.prune as mtp
import torch.multiprocessing as mp
from datasets import load_dataset
from omegaconf import OmegaConf
from pytorch_lightning.trainer.trainer import Trainer
from tqdm import tqdm

from nemo.collections.nlp.models.language_modeling.megatron_gpt_model import MegatronGPTModel
from nemo.collections.nlp.parts.nlp_overrides import NLPDDPStrategy
from nemo.core.config import hydra_runner
from nemo.utils.model_utils import load_config

mp.set_start_method("spawn", force=True)

"""
Nemo pruning example script.

Please consult examples/nlp/language_modeling/conf/megatron_gpt_prune.yaml config on available pruning arguments,
models supported as well as how to set up data and inference for calibration (with defaults recommended).

Example usage:
```
python examples/nlp/language_modeling/megatron_gpt_prune.py \
    model.restore_from_path=llama3.1-8b-base.nemo \
    model.tensor_model_parallel_size=1 \
    model.pipeline_model_parallel_size=8 \
    trainer.num_nodes=1 \
    trainer.precision=bf16 \
    trainer.devices=8 \
    prune.save_path=llama3.1-8b-base-pruned.nemo \
    prune.ffn_hidden_size=3584 \
    prune.num_attention_heads=8 \
    prune.num_query_groups=4
```
where tensor_model_parallel_size must be 1 because of the current prune API limitation
"""


def get_calib_data_iter(data="cnn_dailymail", batch_size=64, calib_size=512, max_sequence_length=512):
    if data == "wikitext":
        dataset = load_dataset("wikitext", "wikitext-103-v1", split="train")
        text_column = "text"
    elif data == "cnn_dailymail":
        dataset = load_dataset("cnn_dailymail", name="3.0.0", split="train")
        text_column = "article"
    else:
        # Assume a local JSON dataset with a column named "text"
        dataset = load_dataset("json", data_files=data, split="train")
        text_column = "text"
    calib_size = max(min(len(dataset), calib_size), batch_size)
    for i in range(calib_size // batch_size):
        batch = dataset[i * batch_size : (i + 1) * batch_size][text_column]
        for j in range(len(batch)):
            batch[j] = batch[j][:max_sequence_length]
        yield batch


@hydra_runner(config_path="conf", config_name="megatron_gpt_prune")
def main(cfg) -> None:
    # Overwrite model config with the one from the model checkpoint and apply pruning modifications
    model_cfg = load_config(cfg.model.restore_from_path)
    model_cfg.update(cfg.model)
    model_cfg.name = "modelopt"  # Use modelopt transformer spec for pruning

    assert cfg.model.tensor_model_parallel_size == 1, "Pruning currently only supports tensor_model_parallel_size=1"
    assert not cfg.model.sequence_parallel, "Pruning currently does not support sequence parallelism"

    trainer = Trainer(strategy=NLPDDPStrategy(), **cfg.trainer)
    model = MegatronGPTModel.restore_from(
        restore_path=cfg.model.restore_from_path, override_config_path=model_cfg, trainer=trainer
    )

    data_iter = get_calib_data_iter(
        cfg.prune.calib_dataset,
        cfg.inference.batch_size,
        cfg.prune.num_calib_size,
        cfg.inference.max_context_length,
    )
    dataloader = [data for data in data_iter]

    def forward_loop(model):
        # NOTE: Alternatively you can also use `model.forward_bwd_step(data_iter, forward_only=True)`
        # if your model is setup for training.
        model.set_inference_config(OmegaConf.to_container(cfg.inference))
        for i, batch in enumerate(tqdm(dataloader, desc="Calibrating")):
            model.predict_step(batch, i)

    model_pruned, _ = mtp.prune(
        model,
        mode="mcore_gpt_minitron",
        constraints={
            "export_config": {
                k: cfg.prune.get(k)
                for k in ["ffn_hidden_size", "num_attention_heads", "num_query_groups"]
                if cfg.prune.get(k) is not None
            },
        },
        dummy_input=None,  # Not used
        config={"forward_loop": forward_loop},
    )

    model_pruned.save_to(cfg.prune.save_path)


if __name__ == '__main__':
    main()
````